torch/csrc/jit/codegen/cuda/contiguity.cpp

```diff
@@ -512,17 +520,27 @@ void ContigIDs::handle(Merge* merge) {
 
   VectorOfUniqueEntries<IterDomain*> root_ids = root_ids_it->second;
 
+  bool is_indexing_pass = !ignore_consistent_ordering_;
+
   IterDomain* last_root = nullptr;
   for (auto root_id_i : c10::irange(root_domain_.size())) {
     auto root_id = root_domain_[root_id_i];
     if (root_ids.has(root_id)) {
       // ID found, remove it
       root_ids.erase(root_id);
-      // If the last id isn't contiguous that's fine, we can use the stride of
-      // the last iter domain to multiply the contig index.
-      if (!root_contiguity_[root_id_i] && !root_ids.empty()) {
-        // Otherwise this merge->out() isn't a contiguously indexable ID
-        return;
+      // If we're indexing:
+      // we could still potentially consider this ID linearly indexable, as we
+      // could multiple the index by the last root's stride.
+      //
+      // If we're computing predicates (ignore_consistent_ordering_==true),
+      // then we don't have this same constraint, we can just ignore
+      // contiguity of the roots all together.
```

Comment:
Makes sense. Thanks for clarification.

Reply:
Actually, this isn't correct. Even for predicates, some root domains are marked as non-contiguous when they need to be predicated at root domains.

Some cleanups would be needed here. I'll work on it.

Reply:
A problem with the design is `root_contiguity` is used both as indicating contiguity AND if domains are not necessary to be predicated at root. These are two separate concepts, so I split out the latter as `final_ids`.